### PR TITLE
Countdown Timer

### PR DESCRIPTION
**Pull request by Crystalwarrior**
(Remake of PR https://github.com/AttorneyOnline/AO2-Client/pull/266 but able to be contributed to)

An incredibly powerful asset in any GM's hands, this is useful for a huge number of role-plays: Danganronpa Class Trials, Nonary Game countdown, Ambidex Game phases, Main Game phases, etc. etc. etc.
Literally anything that needs a countdown timer will benefit from this feature.

**Absolutely Important**:
 - [x] Countdown Timer label class
 - [x] Configurable from courtroom_config.ini and courtroom_fonts.ini
 - [x] Can be started from the server using a unique packet
 - [x] Can be stopped and resumed from the server using a unique packet
 - [x] Accounts for the packet travel time and ensures that everyone is in sync when the timer starts/stops (ROLLBACK NETCODE 2.9)
 - [x] Can be hidden and unhidden by the server using a unique packet
 - [x] Ability to define more than 1 timer (at least 2)
 - [ ] Define timer background image(s)

**TBD**:
 - [ ] Ability to tell the timer to count *up* towards something
 - [ ] Ability to turn the timer into GTA-Style AM/PM day cycle timer

### CODE REVIEW

File: include/courtroom.h [hunk 1]
```diff
@@ -298,11 +307,16 @@ class Courtroom : public QMainWindow {
 
   QVector<chatlogpiece> ic_chatlog_history;
 
-  // triggers ping_server() every 60 seconds
+  // triggers ping_server() every 1 second
```
Comment (oldmud0) [inline include/courtroom.h:310]:
```suggestion
  // triggers ping_server() every 45 seconds
```

**Review by oldmud0 — COMMENTED** [review]
aw h*ck

**Comment by in1tiate** [comment]
How's theme support for this?

**Comment by Crystalwarrior** [comment]
> 
> 
> How's theme support for this?

its ok, but it needs a "timer background" image, as well as support for multiple indexed timers on the theme level

**Review by stonedDiscord — DISMISSED** [review]

File: src/courtroom.cpp [hunk 2]
```diff
@@ -5066,34 +5072,52 @@ void Courtroom::announce_case(QString title, bool def, bool pro, bool jud,
...
 }
 
-void Courtroom::start_clock()
+void Courtroom::start_clock(int id)
 {
-  ui_clock->start();
+  if (id < max_clocks && ui_clock[id] != nullptr)
+  {
+    ui_clock[id]->start();
+  }
```
Comment (oldmud0) [inline src/courtroom.cpp:5080]: Forgot a check for id < 0. Same goes for the other methods.
```suggestion
  if (id < 0 || id >= max_clocks || ui_clock[id] == nullptr)
    return;

  ui_clock[id]->start();
```

**Review by oldmud0 — COMMENTED** [review]
Looks good. Need to update server, theme, and docs.

File: src/packet_distribution.cpp [hunk 3]
```diff
@@ -574,6 +574,64 @@ void AOApplication::server_packet_received(AOPacket *p_packet)
...
   }
+  else if (header == "TI") { // Timer packet
+    if (!courtroom_constructed || f_contents.size() < 2)
+      goto end;
+
+    // Timer ID is reserved as argument 0
+    int id = f_contents.at(0).toInt();
+
+    // ID is invalid
+    if (id < 0 || id >= w_courtroom->max_clocks)
```
Comment (oldmud0) [inline src/packet_distribution.cpp:585]:
```
src/packet_distribution.cpp:585:38: error: 'max_clocks' is a private member of 'Courtroom'
    if (id < 0 || id >= w_courtroom->max_clocks)
                                     ^
include/courtroom.h:566:20: note: declared private here
  static const int max_clocks = 5;
                   ^
```

**Review by oldmud0 — CHANGES_REQUESTED** [review]
Also fix build error, please.

**Comment by Crystalwarrior** [comment]
Last thing this needs before merging is this:
 - [ ] Define timer background image(s)

File: src/packet_distribution.cpp (hunk 3 above)
Comment (Crystalwarrior) [reply src/packet_distribution.cpp]: fixed

File: src/courtroom.cpp (hunk 2 above)
Comment (Crystalwarrior) [reply src/courtroom.cpp]: fixed

**Comment by oldmud0** [comment]
> [ ]  Define timer background image(s)

Can't you just do that from QSS though?

**Comment by Crystalwarrior** [comment]
> > [ ]  Define timer background image(s)
> 
> Can't you just do that from QSS though?

That's inconsistent, you can define a music display bg in the ini, people will just assume the timers don't have bgs

**Comment by oldmud0** [comment]
Then just remove music display bg from the ini. Ez

**Comment by Crystalwarrior** [comment]
> Then just remove music display bg from the ini. Ez

Then we break existing themes. EZ EZ

**Comment by oldmud0** [comment]
> > Then just remove music display bg from the ini. Ez
> 
> Then we break existing themes. EZ EZ

They're all gonna be broken anyway by the time we're done

**Comment by Crystalwarrior** [comment]
> 
> 
> > > Then just remove music display bg from the ini. Ez
> > 
> > 
> > Then we break existing themes. EZ EZ
> 
> They're all gonna be broken anyway by the time we're done

only in ao3, 2.9 still compatible baby

**Review by oldmud0 — DISMISSED** [review]

**Review by oldmud0 — APPROVED** [review]

**Pull request merged**